File: apps/web/ios/App/App/CarLibrary.swift
import Foundation

/// The library, as the car screen sees it.
///
/// Connecting to CarPlay can launch the app with no window scene at all: there
/// is no WebView, no React, no server session, and nothing to ask what is on
/// the shelf. So the app keeps a snapshot of the library on disk, written by
/// the web layer whenever it changes, and the car screen reads that. Every
/// value the car needs to start playing — track URLs with their media token,
/// the saved position, the book's gain — is in the snapshot, because none of it
/// can be fetched from a cold launch parked in a garage.
struct CarLibrarySnapshot: Codable {
    /// `serverStorageKey:userId`, the first two thirds of the recovery scope
    /// key the web player uses. Kept whole rather than rebuilt natively so the
    /// two sides cannot drift apart.
    var scopePrefix: String
    var playbackRate: Double
    var updatedAt: Double
    var books: [CarLibraryBook]

    static let empty = CarLibrarySnapshot(scopePrefix: "", playbackRate: 1, updatedAt: 0, books: [])

    func recoveryScopeKey(forBookId bookId: String) -> String {
        "\(scopePrefix):\(bookId)"
    }

    func book(withId bookId: String) -> CarLibraryBook? {
        books.first { $0.id == bookId }
    }

    func resuming(_ book: CarLibraryBook, sessions: [CarPlaybackSession]) -> CarLibraryBook {
        guard let session = sessions.filter({
            $0.bookId == book.id && $0.updatedAt > updatedAt && $0.bookPositionSeconds.isFinite
        }).max(by: { $0.updatedAt < $1.updatedAt }) else { return book }
        var resumed = book
        resumed.positionSeconds = session.bookPositionSeconds
        resumed.status = session.finished ? "finished" : "inProgress"
        return resumed
    }

    func libraryGroups(maximumItemCount: Int) -> [(String, [CarLibraryBook])] {
        var remaining = max(0, maximumItemCount)
        return [
            ("Reading", books.filter { $0.isInProgress }),
            ("Not started", books.filter { !$0.isInProgress && !$0.isFinished }),
            ("Finished", books.filter { $0.isFinished })
        ].compactMap { title, books in
            let items = Array(books.prefix(remaining))
            remaining -= items.count
            return items.isEmpty ? nil : (title, items)
        }
    }
}

struct CarLibraryBook: Codable {
    var id: String
    var title: String
    var author: String?
    var artworkUrl: String?
    var durationSeconds: Double?
    /// Whole-book position, in seconds, as the web layer last knew it.
    var positionSeconds: Double
    var status: String
    var downloaded: Bool
    var volumeGain: Double?
    var tracks: [CarLibraryTrack]
    var chapters: [CarLibraryChapter]

    var isFinished: Bool { status == "finished" }
    var isInProgress: Bool { status == "inProgress" }

    /// Where a plain "play this book" starts. A book that was played to the
    /// end starts over rather than resuming a second past the last word.
    var resumePositionSeconds: Double {
        if isFinished { return 0 }
        if let durationSeconds, durationSeconds > 0, positionSeconds >= durationSeconds - 1 { return 0 }
        return max(0, positionSeconds)
    }

    /// Where in the queue a whole-book position lands: the track holding it,
    /// and the offset into that track.
    func target(atBookPosition position: Double) -> (trackIndex: Int, positionInTrack: Double)? {
        guard !tracks.isEmpty else { return nil }
        let position = max(0, position)
        for (index, track) in tracks.enumerated().reversed() where position >= track.bookOffsetSeconds {
            return (index, position - track.bookOffsetSeconds)
        }
        return (0, position)
    }

    var percentComplete: Double? {
        guard let durationSeconds, durationSeconds > 0 else { return nil }
        return min(1, max(0, positionSeconds / durationSeconds))
    }
}

struct CarLibraryTrack: Codable {
    var id: String
    var title: String
    var url: String
    var durationSeconds: Double?
    var bookOffsetSeconds: Double
}

struct CarLibraryChapter: Codable {
    var title: String
    var startSeconds: Double
    var durationSeconds: Double
    var trackId: String?
}

/// Progress made in the car, waiting for the web layer to save it.
///
/// The car never writes to the server itself. Every defense against a bad
/// progress write — the staleness check, the suspect-reset rule, the
/// book-summary fallback — lives in the web player and the server, and a second
/// writer that skipped them is exactly the failure those defenses exist for. So
/// a car session leaves its position here, and the app persists it through the
/// usual path the next time it runs.
struct CarPlaybackSession: Codable {
    var bookId: String
    var trackId: String
    var positionSeconds: Double
    var bookPositionSeconds: Double
    var durationSeconds: Double?
    var updatedAt: Double
    var finished: Bool
    /// The listener jumped rather than simply listening on. The server refuses
    /// a write that drops a substantial position back to near zero unless the
    /// client says the move was deliberate, so the car has to pass that intent
    /// along with the position.
    var intentionalRegression: Bool
}

extension CarLibraryBook {
    /// The line under a book's title on the car screen: who wrote it, and what
    /// tapping it will do.
    var carSubtitle: String {
        var parts: [String] = []
        if let author, !author.isEmpty { parts.append(author) }
        if isFinished {
            parts.append("Finished")
        } else if let durationSeconds, durationSeconds > 0 {
            parts.append(
                resumePositionSeconds > 0
                    ? "\(carDurationLabel(seconds: durationSeconds - resumePositionSeconds)) left"
                    : carDurationLabel(seconds: durationSeconds)
            )
        }
        return parts.joined(separator: " · ")
    }
}

/// How much listening is left, in the shelf's own phrasing.
func carDurationLabel(seconds: Double) -> String {
    let total = Int(max(0, seconds.rounded()))
    let hours = total / 3600
    let minutes = (total % 3600) / 60
    if hours > 0 { return minutes > 0 ? "\(hours)h \(minutes)m" : "\(hours)h" }
    if minutes > 0 { return "\(minutes)m" }
    return "under a minute"
}

/// A position in the book, as a clock reads it. The duration label answers
/// "how much is left"; a chapter list answers "where does this start", and
/// "under a minute" is not an answer to that question.
func carTimestampLabel(seconds: Double) -> String {
    let total = Int(max(0, seconds.rounded()))
    let hours = total / 3600, minutes = (total % 3600) / 60, remainder = total % 60
    return hours > 0
        ? String(format: "%d:%02d:%02d", hours, minutes, remainder)
        : String(format: "%d:%02d", minutes, remainder)
}

File: apps/web/src/carLibrary.ts
import { buildChapterSegments } from "./chapters.ts";
import type { Book, Track } from "./types.ts";

/** Stops late snapshot work from republishing a library after sign-out. */
export class CarLibraryAccess {
  private scope: string | null = null;

  begin(scope: string) { this.scope = scope; }
  end() { this.scope = null; }
  allows(scope: string) { return this.scope !== null && this.scope === scope; }
}

/**
 * The library as the car sees it, and the progress a drive leaves behind.
 *
 * The car screen cannot run the app: connecting to a car may launch the process
 * with a CarPlay scene and nothing else — no WebView, no session, no network.
 * So the app writes this snapshot to the native side whenever the library
 * changes, and the car reads it from disk.
 *
 * Progress travels the other way. The car deliberately writes nothing itself;
 * every guard against a bad position — the staleness check, the suspect-reset
 * rule, the retry before reconciling — lives in the app and the server, and a
 * second writer that skipped them is the exact failure those guards exist for.
 * A drive therefore leaves a session record behind, and the app saves it
 * through its usual path the next time it runs.
 */

export type CarLibraryTrack = {
  id: string;
  title: string;
  url: string;
  durationSeconds?: number;
  bookOffsetSeconds: number;
};

export type CarLibraryChapter = {
  title: string;
  startSeconds: number;
  durationSeconds: number;
  trackId?: string;
};

export type CarLibraryBook = {
  id: string;
  title: string;
  author?: string;
  artworkUrl?: string;
  durationSeconds?: number;
  positionSeconds: number;
  status: "notStarted" | "inProgress" | "finished";
  downloaded: boolean;
  volumeGain?: number;
  tracks: CarLibraryTrack[];
  chapters: CarLibraryChapter[];
};

export type CarLibrarySnapshot = {
  /** `serverStorageKey:userId` — the recovery scope key without its book. */
  scopePrefix: string;
  playbackRate: number;
  updatedAt: number;
  books: CarLibraryBook[];
};

export type CarPlaybackSession = {
  bookId: string;
  trackId: string;
  positionSeconds: number;
  bookPositionSeconds: number;
  durationSeconds?: number;
  updatedAt: number;
  finished: boolean;
  /** The listener jumped rather than listened on; the server needs to be told. */
  intentionalRegression: boolean;
};

/** Tolerates anything the bridge hands back rather than losing a whole drive. */
export function parseCarSessions(raw: unknown): CarPlaybackSession[] {
  if (typeof raw !== "string" || raw.length === 0) return [];
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    return [];
  }
  if (!Array.isArray(parsed)) return [];
  return parsed.filter((entry): entry is CarPlaybackSession =>
    !!entry
    && typeof entry === "object"
    && typeof (entry as CarPlaybackSession).bookId === "string"
    && typeof (entry as CarPlaybackSession).trackId === "string"
    && Number.isFinite((entry as CarPlaybackSession).bookPositionSeconds)
    && Number.isFinite((entry as CarPlaybackSession).updatedAt));
}

/**
 * What the car should be told about the library.
 *
 * `resolveTrackUrl` is injected because a downloaded book plays from a file on
 * the device and a streamed one from the server with a media token, and only
 * the caller can tell which is which.
 */
export async function buildCarLibrarySnapshot(options: {
  scopePrefix: string;
  playbackRate: number;
  books: Book[];
  downloadedBookIds: ReadonlySet<string>;
  bookGains: Readonly<Record<string, number>>;
  coverUrl: (book: Book) => Promise<string | null>;
  resolveTrackUrl: (book: Book, track: Track) => Promise<string | null>;
  now?: () => number;
}): Promise<CarLibrarySnapshot> {
  const books = await Promise.all(
    options.books.map((book) => carBook(book, options))
  );
  return {
    scopePrefix: options.scopePrefix,
    playbackRate: options.playbackRate,
    updatedAt: (options.now ?? Date.now)(),
    // A book with nothing playable would be a dead end on the car screen.
    books: books.filter((book): book is CarLibraryBook => book !== null)
  };
}

async function carBook(
  book: Book,
  options: {
    downloadedBookIds: ReadonlySet<string>;
    bookGains: Readonly<Record<string, number>>;
    coverUrl: (book: Book) => Promise<string | null>;
    resolveTrackUrl: (book: Book, track: Track) => Promise<string | null>;
  }
): Promise<CarLibraryBook | null> {
  const downloaded = options.downloadedBookIds.has(book.id) || book.source === "device";
  const resolved = await Promise.all(
    book.tracks.map(async (track, index) => {
      const url = await options.resolveTrackUrl(book, track).catch(() => null);
      return url === null ? null : {
        id: track.id,
        title: track.title,
        url,
        ...(Number.isFinite(track.durationSeconds ?? NaN)
          ? { durationSeconds: track.durationSeconds as number }
          : {}),
        bookOffsetSeconds: trackOffsetSeconds(book, index)
      } satisfies CarLibraryTrack;
    })
  );
  const tracks = resolved.filter((track): track is CarLibraryTrack => track !== null);
  if (tracks.length === 0) return null;

  const durationSeconds = book.durationSeconds
    ?? book.progress?.durationSeconds
    ?? durationFromTracks(book);
  const chapters = buildChapterSegments(book.chapters, durationSeconds ?? 0).map((chapter) => ({
    title: chapter.title,
    startSeconds: chapter.startSeconds,
    durationSeconds: chapter.durationSeconds,
    trackId: chapter.trackId
  }));
  const status = book.progress?.status;
  const artworkUrl = await options.coverUrl(book).catch(() => null);
  return {
    id: book.id,
    title: book.title,
    ...(book.author ? { author: book.author } : {}),
    ...(artworkUrl ? { artworkUrl } : {}),
    ...(durationSeconds && durationSeconds > 0 ? { durationSeconds } : {}),
    positionSeconds: Math.max(0, book.progress?.bookPositionSeconds ?? 0),
    status: status === "inProgress" || status === "finished" ? status : "notStarted",
    downloaded,
    ...(Number.isFinite(options.bookGains[book.id] ?? NaN)
      ? { volumeGain: options.bookGains[book.id] }
      : {}),
    tracks,
    chapters
  };
}

/**
 * A track's offset into the whole book. The car needs it for the same reason
 * the player does: the position it stores is a whole-book one, and every track
 * is played from its own zero.
 */
function trackOffsetSeconds(book: Book, trackIndex: number) {
  return book.tracks
    .slice(0, Math.max(0, trackIndex))
    .reduce((sum, track) => sum + (track.durationSeconds ?? 0), 0);
}

function durationFromTracks(book: Book) {
  const total = book.tracks.reduce((sum, track) => sum + (track.durationSeconds ?? 0), 0);
  return total > 0 ? total : null;
}

/**
 * Whether a car session is worth saving over what the app already has.
 *
 * A drive that ended where the book already was writes nothing: the position is
 * unchanged, and a redundant write is one more chance to get progress wrong.
 */
export function carSessionIsWorthSaving(
  session: CarPlaybackSession,
  book: Pick<Book, "progress">
): boolean {
  if (!Number.isFinite(session.bookPositionSeconds)) return false;
  const known = book.progress?.bookPositionSeconds ?? 0;
  return Math.abs(session.bookPositionSeconds - known) >= 1;
}
